sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DefaultSynchronousMetricStorage.java

```diff
@@ -169,13 +181,12 @@ public MetricData collect(
     if (reset) {
       AggregatorHolder<T, U> holder = this.aggregatorHolder;
       this.aggregatorHolder = new AggregatorHolder<>();
-      Lock writeLock = holder.lock.writeLock();
-      writeLock.lock();
-      try {
-        aggregatorHandles = holder.aggregatorHandles;
-      } finally {
-        writeLock.unlock();
-      }
+      // Set recordsInProgress to -1, which causes record operations to re-read the volatile
+      // this.aggregatorHolder
+      // Use compareAndSet in a loop to confirm we only set recordsInProgress if there aren't any
+      // records operations in progress.
+      while (!holder.recordsInProgress.compareAndSet(0, -1)) {}
+      aggregatorHandles = holder.aggregatorHandles;
```

Comment:
what if the collect thread stalls in between the two above lines (188 and 189)?

doesn't that effectively block the record threads?

Reply:
Once we pass line 188, we've guaranteed that no more record threads are interacting with the instance of `AggregatorHolder` needed for the current collection. New record operations operate against the newly set `AggregatorHolder`, which was set previously on line 183.

Reply:
I think I understand now